### PR TITLE
Port modules collection script to golang

### LINKED ISSUE #9130: Port over modules collection script to golang
We are trying to [move away](https://github.com/elastic/beats/pull/9118#discussion_r234195444) from Python scripts called from Makefiles to golang scripts called from magefiles. Currently the [modules collection script in Metricbeat](https://github.com/elastic/beats/blob/master/metricbeat/scripts/modules_collector.py) is written in Python. It should be ported over to a golang script.

### PR DESCRIPTION
**Pull request by eloyekunle**
Closes elastic/beats#9130

### CODE REVIEW

File: metricbeat/Makefile [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # Name can be overwritten, as Metricbeat is also a library
-BEAT_NAME?=metricbeat
+export BEAT_NAME?=metricbeat
```
Comment (jsoriano) [inline metricbeat/Makefile]: Why is this change needed?

File: metricbeat/Makefile [hunk 2]
```diff
@@ -9,7 +9,7 @@ ES_BEATS?=..
 GOX_OS=netbsd linux windows
 GOX_FLAGS=-arch="amd64 386 arm ppc64 ppc64le"
 
-DOCS_BRANCH=$(shell grep doc-branch ../libbeat/docs/version.asciidoc | cut -c 14-)
+export DOCS_BRANCH=$(shell grep doc-branch ../libbeat/docs/version.asciidoc | cut -c 14-)
```
Comment (jsoriano) [inline metricbeat/Makefile]: Why is this change needed?

File: metricbeat/docs/modules/aws/ebs.asciidoc [hunk 3]
```diff
@@ -9,7 +9,6 @@ beta[]
 
 include::../../../../x-pack/metricbeat/module/aws/ebs/_meta/docs.asciidoc[]
 
-This is a default metricset. If the host module is unconfigured, this metricset is enabled by default.
```
Comment (jsoriano) [inline metricbeat/docs/modules/aws/ebs.asciidoc]: This shouldn't change, try running `BEAT_STRICT_PERMS=false make update`.

File: metricbeat/scripts/mage/modules_collector.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+# Docs: https://www.elastic.co/guide/en/beats/%[2]s/%[3]s/%[2]s-module-%[1]s.html
+
+`
+	r, err := regexp.Compile(`.+\.reference\.yml`)
+	if err != nil {
+		return err
+	}
+
+	beatName := os.Getenv("BEAT_NAME")
+	docsBranch := os.Getenv("DOCS_BRANCH")
```
Comment (jsoriano) [inline metricbeat/scripts/mage/modules_collector.go]: I think you can get beat name and docs branch from `github.com/elastic/beats/dev-tools/mage`, as `BeatName` and `BeatDocBranch()`.

File: metricbeat/scripts/mage/modules_collector.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+	path, err := filepath.Abs("module")
```
Comment (jsoriano) [inline metricbeat/scripts/mage/modules_collector.go]: Use helpers for paths in `github.com/elastic/beats/dev-tools/mage`, like `OSSBeatDir("module")`.

File: script/modules_collector.py [hunk 6]
```diff
@@ -1,61 +0,0 @@
...
-    parser = argparse.ArgumentParser(
-        description="Collects modules confs")
-    parser.add_argument("--beat", help="Beat name")
-    parser.add_argument("--docs_branch", help="Docs branch")
-
-    args = parser.parse_args()
-    beat_name = args.beat
-    docs_branch = args.docs_branch
-
-    collect(beat_name, docs_branch)
```
Comment (jsoriano) [inline script/modules_collector.py]: :tada:

File: metricbeat/scripts/mage/modules_collector.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+	if err != nil {
+		return err
+	}
+
+	modules, err := ioutil.ReadDir("module")
+	if err != nil {
+		return err
+	}
+
+	if err = os.Mkdir("modules.d", os.ModeDir|os.ModePerm); err != nil {
```
Comment (jsoriano) [inline metricbeat/scripts/mage/modules_collector.go]: Also here consider using helpers in `github.com/elastic/beats/dev-tools/mage`.

File: metricbeat/Makefile (hunk 1 above)
Comment (eloyekunle) [reply metricbeat/Makefile]: Added it so the Magefile could pick it out with `getEnv`, but unnecessary with the helpers.

File: metricbeat/scripts/mage/modules_collector.go (hunk 4 above)
Comment (eloyekunle) [reply metricbeat/scripts/mage/modules_collector.go]: done

File: metricbeat/scripts/mage/modules_collector.go (hunk 5 above)
Comment (eloyekunle) [reply metricbeat/scripts/mage/modules_collector.go]: done

File: metricbeat/scripts/mage/modules_collector.go (hunk 7 above)
Comment (eloyekunle) [reply metricbeat/scripts/mage/modules_collector.go]: The closest helper I could find is `CreateDir`, but it expects a full file path.

File: metricbeat/Makefile [hunk 8]
```diff
@@ -34,7 +34,7 @@ configs: python-env
 	@cat ${ES_BEATS}/metricbeat/_meta/common.reference.yml > _meta/beat.reference.yml
 	@${PYTHON_ENV}/bin/python ${ES_BEATS}/script/config_collector.py --beat ${BEAT_NAME} --full $(PWD) >> _meta/beat.reference.yml
 	@rm -rf modules.d
-	${PYTHON_ENV}/bin/python ${ES_BEATS}/script/modules_collector.py --beat ${BEAT_NAME} --docs_branch=$(DOCS_BRANCH)
+	mage ModulesCollector
```
Comment (jsoriano) [inline metricbeat/Makefile]: Umm, I have just found that we already have a target that could do this work
```suggestion
	mage config
```

File: metricbeat/scripts/mage/modules_collector.go [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	devtools "github.com/elastic/beats/dev-tools/mage"
+)
+
+// CollectModules collects module configs to modules.d
+func CollectModules() error {
```
Comment (jsoriano) [inline metricbeat/scripts/mage/modules_collector.go]: I think this is implementing something similar to `metricbeat.GenerateDirModulesD` (if not the same thing)

File: metricbeat/Makefile (hunk 8 above)
Comment (eloyekunle) [reply metricbeat/Makefile]: Yup, it works.

**Comment by eloyekunle** [comment]
@jsoriano You're right, the targets work fine. I've removed the new script I added.

File: metricbeat/magefile.go [hunk 10]
```diff
@@ -73,6 +73,11 @@ func CrossBuildGoDaemon() error {
 	return devtools.CrossBuildGoDaemon()
 }
 
+// ModulesCollector collects module configs to modules.d
+func ModulesCollector() error {
+	return metricbeat.CollectModules()
+}
+
```
Comment (jsoriano) [inline metricbeat/magefile.go]: I guess this can be removed too? CI builds are failing because of that.

Comment (eloyekunle) [reply metricbeat/magefile.go]: Done, removed.

**Comment by jsoriano** [comment]
ok to test

**Comment by eloyekunle** [comment]
the failing tests are not related to this PR

**Comment by jsoriano** [comment]
@eloyekunle I think that at least [this failure](https://travis-ci.org/elastic/beats/jobs/581325350#L766) is related to this change. There may be more things affected for the change in this make target.

**Comment by eloyekunle** [comment]
@jsoriano I've fixed the Travis build. I don't know why `beats-ci` is failing.

**Comment by jsoriano** [comment]
Thanks @eloyekunle! Don't worry about the failure in Jenkins, this is not related.

Code-wise this PR looks good I will test this a little bit locally before approving and merging.

File: metricbeat/Makefile [hunk 11]
```diff
@@ -34,8 +34,8 @@ configs: python-env
 	@cat ${ES_BEATS}/metricbeat/_meta/common.reference.yml > _meta/beat.reference.yml
 	@${PYTHON_ENV}/bin/python ${ES_BEATS}/script/config_collector.py --beat ${BEAT_NAME} --full $(PWD) >> _meta/beat.reference.yml
 	@rm -rf modules.d
-	${PYTHON_ENV}/bin/python ${ES_BEATS}/script/modules_collector.py --beat ${BEAT_NAME} --docs_branch=$(DOCS_BRANCH)
-	@chmod go-w modules.d/*
+	mage config
+	@chmod go-w ${ES_BEATS}/metricbeat/modules.d/*
```
Comment (jsoriano) [inline metricbeat/Makefile]: @eloyekunle I have been testing this branch, and it doesn't work well for generators, the change in this chmod shouldn't be needed, this `modules.d` directory should be generated by `make configs` also in generated beats.

I think that we should change the magefile in `generator/metricbeat/{beat}/magefile.go` so it does something similar of what mage does for the real Metricbeat.

It would be to rename the `Config()` function to something like `configYML()` and then implement `Config()` as:
```
func Config() {
        mg.Deps(configYML, metricbeat.GenerateDirModulesD)
}
```

Could you revert this change to the chmod and give a try to this proposal?

Comment (eloyekunle) [reply metricbeat/Makefile]: @jsoriano I have made the change to `generator/metricbeat/{beat}/magefile.go`.

**Comment by jsoriano** [comment]
@fearful-symmetry could you also take a look to this change? I would like to have a second opinion in case I am mising something.

**Comment by jsoriano** [comment]
jenkins, test this again please

**Comment by fearful-symmetry** [comment]
Yah, I'm lost on that too. uhhh...rebase?

**Comment by jsoriano** [comment]
@fearful-symmetry you are right, this PR fails with master, thanks!

@eloyekunle could you update this branch with master? You will also need to add this import to `generator/metricbeat/{beat}/magefile.go`:
```
metricbeat "github.com/elastic/beats/metricbeat/scripts/mage"
```

**Comment by eloyekunle** [comment]
@jsoriano I've rebased, and added the import. Thanks :+1:

**Comment by jsoriano** [comment]
@eloyekunle thanks! this looks good now, merging!
